fw/http_frame.c

```diff
@@ -1650,36 +1660,65 @@ tfw_h2_frame_recv(void *data, unsigned char *buf, unsigned int len,
 	}
 
 	T_FSM_STATE(HTTP2_RECV_DATA) {
-		FRAME_FSM_READ(ctx->to_read);
+		FRAME_FSM_READ_PYLD();
 
 		if (tfw_h2_stream_state_process(ctx))
 			FRAME_FSM_EXIT(T_DROP);
 
+		if (unlikely(ctx->to_read)) {
+			ctx->state = HTTP2_RECV_APP_DATA_POST;
+			FRAME_FSM_NEXT();
+		}
+
 		FRAME_FSM_EXIT(T_OK);
 	}
 
 	T_FSM_STATE(HTTP2_RECV_HEADER) {
-		FRAME_FSM_READ(ctx->to_read);
+		FRAME_FSM_READ_PYLD();
 
 		if (tfw_h2_headers_process(ctx))
 			FRAME_FSM_EXIT(T_DROP);
 
+		if (unlikely(ctx->to_read)) {
+			ctx->state = HTTP2_RECV_APP_DATA_POST;
+			FRAME_FSM_NEXT();
+		}
+
 		FRAME_FSM_EXIT(T_OK);
 	}
 
 	T_FSM_STATE(HTTP2_RECV_CONT) {
-		FRAME_FSM_READ(ctx->to_read);
+		FRAME_FSM_READ_PYLD();
 
 		if (tfw_h2_stream_state_process(ctx))
 			FRAME_FSM_EXIT(T_DROP);
 
+		if (unlikely(ctx->to_read)) {
+			ctx->state = HTTP2_RECV_APP_DATA_POST;
+			FRAME_FSM_NEXT();
```

Comment:
Why do we reenter the `switch` statement, why not just to directly jump to the state?